### PR TITLE
Add tests for edge cases of multiple membership changes in sync period

### PR DESCRIPTION
**Pull request by anoadragon453**
Addresses https://github.com/element-hq/synapse/issues/14683 case seen in the real world.

Requires https://github.com/element-hq/synapse/pull/18648 for these tests to pass in Synapse.

### Pull Request Checklist

- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)

### CODE REVIEW

File: tests/csapi/sync_test.go [hunk 1]
```diff
@@ -731,6 +731,145 @@ func TestRoomSummary(t *testing.T) {
...
+	)
+
+	// Wait until Bob is kicked.
+	alice.MustSyncUntil(t, client.SyncReq{}, client.SyncLeftFrom(bob.UserID, roomID))
+
+	// Alice re-invites Bob
+	alice.MustInviteRoom(t, roomID, bob.UserID)
+
+	// 4. Bob does a sync
+	jsonRes, _ := bob.MustSync(t, client.SyncReq{Since: bobSince})
```
Comment (MadLittleMods) [inline tests/csapi/sync_test.go:860]: This doesn't guarantee that the leave/invite will be federated to Bob's homeserver by the time Bob syncs.

Perhaps we need another `hs2` user in the room to observe first

If we just use `bob.MustSyncUntil(t, client.SyncReq{Since: bobSince}, client.SyncInvitedTo(bob.UserID, roomID))` the leave and invite could be separated in different sync responses as we paginate.

**Review by MadLittleMods — CHANGES_REQUESTED** [review]

File: tests/csapi/sync_test.go (hunk 1 above)
Comment (anoadragon453) [reply tests/csapi/sync_test.go:860]: Very good point. Another user would work. I'll modify the test to include that.

File: tests/csapi/sync_test.go [hunk 2]
```diff
@@ -731,6 +731,163 @@ func TestRoomSummary(t *testing.T) {
...
+	// For kicking, we need to use the POST /rooms/{roomId}/kick endpoint
+	alice.MustDo(t, "POST", []string{"_matrix", "client", "v3", "rooms", roomID, "kick"},
+		client.WithJSONBody(t, map[string]any{
+			"user_id": bob.UserID,
+		}),
+	)
+
+	// Wait until Bob is kicked.
+	aliceSince := alice.MustSyncUntil(t, client.SyncReq{}, client.SyncLeftFrom(bob.UserID, roomID))
+	charlieSince = charlie.MustSyncUntil(t, client.SyncReq{Since: charlieSince}, client.SyncLeftFrom(bob.UserID, roomID))
```
Comment (MadLittleMods) [inline tests/csapi/sync_test.go:865]: I don't think we need to worry about getting a `since` token in any of these places execpt for `bob` which we are actually testing incremental sync for.

The other sync's are just asserting pre-conditions.

```suggestion
	charlie.MustSyncUntil(t, client.SyncReq{}, client.SyncLeftFrom(bob.UserID, roomID))
```

Comment (MadLittleMods) [inline tests/csapi/sync_test.go:865]: Add a comment about why we wait here as well

(needs wrapping)

```suggestion
	// Wait until hs2 sees that Bob has been kicked. If we don't wait, then we have no idea whether `hs2` received the kick event.
	charlieSince = charlie.MustSyncUntil(t, client.SyncReq{Since: charlieSince}, client.SyncLeftFrom(bob.UserID, roomID))
```
